src/buildercore/keypair.py

```diff
@@ -17,22 +16,21 @@
 def s3_keypair_key(stackname):
     return config.KEYPAIR_PREFIX + stackname + ".pem"
 
-@if_enabled('write-keypairs-to-s3', silent=True)
 def write_keypair_to_s3(stackname):
     # this is the path to where .save() puts it
     # http://boto.readthedocs.io/en/latest/ref/ec2.html#boto.ec2.keypair.KeyPair
     path = stack_pem(stackname, die_if_doesnt_exist=True)
     key = s3_keypair_key(stackname)
-    s3.write(key, open(path, 'r'))
+    with open(path, 'r') as fp:
+        pem_contents = fp.read()
```

Comment:
a simple `pem_contents = open(path, 'r').read()` is all that is needed. the file handle is closed immediately

Reply:
it doesn't seem closed after these two calls?
```
>>> fp = open('projects/elife.yaml')
>>> len(fp.read())
77047
>>> fp.seek(0)
>>> len(fp.read())
77047
```

Reply:
well you're keeping a reference to it with `fp` ;)

After some googling, what I recall is that the CPython garbage collector closes file handles with no references to it immediately. Given that explicit is better than implicit and all that and we shouldn't rely on compiler/interpreter/garbage collector behaviour, I don't really mind.